crypto/bls/signature_batch.go

```diff
@@ -32,10 +32,39 @@ func NewSet() *SignatureBatch {
 
 // Join merges the provided signature batch to out current one.
 func (s *SignatureBatch) Join(set *SignatureBatch) *SignatureBatch {
+	if cap(s.Signatures)-len(s.Signatures) < len(set.Signatures) &&
+		cap(set.Signatures)-len(set.Signatures) >= len(s.Signatures) {
+		return set.Join(s)
+	}
+
+	total := len(s.Signatures) + len(set.Signatures)
+
+	if cap(s.Signatures) < total {
```

Comment:
re this updated version - to get the amortization benefit you want to straight up append, which will let the go runtime do its default behavior of allocating `cap()*2`. So if you want to go this way, you can just get rid of lines 40-61.